### PR TITLE
Multi-type and nestable parameters

### PR DESCRIPTION
**Pull request by koosemose**
As discussed in #532 and #302. While there is still room for refinement, particularly as things change around it (particularly any changes to how Xml is handled), this allows more complex custom parameters, which will aid in modding and limiting the number of variables need to be added to the furniture class (and any other class which is changed to make use of this).

While I don't like including multiple changes in one PR, the changes to furniture.xml will cause errors if applied without the changes to the furniture class and the new Parameter system will cause errors in Furniture.lua, if it isn't updated.

Summary: Allows Params to be nested and directly store strings (which can be read as Floats when needed), such as this example from furniture.xml:

```
        <Params>
            <Param name="gas_limit" value="0.2" />
            <Param name="gas_per_second" value="0.16" />
            <Param name="gas_gen">
                <Param name="O2">
                    <Param name="gas_limit" value="0.2" />
                </Param>
                <Param name="N2">
                    <Param name="gas_limit" value="0.8" />
                </Param>
            </Param>
        </Params>
```

A Get method has been implemented for furnParameters, the contents are accessed as follows (using O2 generator's structure as an example.):
`if(Parameters.HasContents())`
Singular Parameters
`Parameters["gas_per_second"].ToFloat();`
`Parameters["gas_per_second"].ToString();`
`Parameters["gas_gen"]["O2"]["gas_limit"].ToFloat();`
`Parameters["gas_per_second"].ChangeFloatValue(.2);`
`Parameters["gas_per_second"].SetValue("1");`
`Parameters["gas_per_second"].SetValue(1);`
Group Parameters
`Parameters["gas_gen"].Keys();`
`if (Parameters["gas_gen"].ContainsKey("O2"));`
`Parameters.AddParameter(new Parameter("max_pressure","1"));`
`Parameters.AddParameter(new Parameter("max_pressure",1));`
Each "layer" is a Parameter, and the tree can be descended with iterators (Parameter["key"]), and final value accessed or changed with any of the singular parameter functions. Has Contents can be used if you are uncertain of the tree's structure, to determine if it is singular or group, and Keys() can assist in traversing to the next level down.

### CODE REVIEW

**Comment by FilippoLeon** [comment]
If you implement parameter as a "property", I think it coul be possible to write:
`furn.parameters["openness"].ToFloat()`
istead of
`furn.GetParameters()["openness"].ToFloat()`

the GetParameters()[], from a functional standpoint, is working, but is screwing up my ocd.

I think there is something wonky with tabs in the switch statement.

**Comment by koosemose** [comment]
My understanding is that there is something weird with accessing Properties from MoonSharp, and that's one of the primary places to access them from, so I had to go with the method version.

**Comment by FilippoLeon** [comment]
I know many do not like it, but "public field"?

**Comment by mikejbrown** [comment]
:+1: I like this, though note that it will likely break any outstanding PRs that touch furniture.

**Comment by koosemose** [comment]
Apparently I was wrong, I could swear I remember in one of the tutorial videos when he was doing the Lua, quill saying there was something wrong with using Properties in Lua. I think for now I'll have both in, just in case it was some other oddity with it I can't remember.

**Comment by koosemose** [comment]
@mikejbrown yeah, I tried to get this done while there were little to no PRs touching anything furniture related, but it's taken so long I think a few managed to sneak in. I'm hoping my little guide to use will ease anyone having to resolve conflicts while using this system, if mine is merged first... otherwise I'm going to have to play catchup again.

**Comment by mikejbrown** [comment]
@koosemose I don't think that mine ( #670 ) conflicts with this _yet_. It would do if I had already started implementing the stockpile accepted inventory stuff as furniture params as I was planning. So I'll hold off on that till this gets merged. (It'll probably be much cleaner with your system anyway.)

**Comment by sboigelot** [comment]
It's ok to conflict with PR about furniture for now I think. We can propose help to other dev too if they need more info about how to fix the lua.

Great work by the way :D

File: Assets/Scripts/Controllers/FurnitureSpriteController.cs [hunk 1]
```diff
@@ -205,17 +205,17 @@ public Sprite GetSpriteForFurniture(Furniture furn)
             // FIXME: All this hardcoding needs to be generalized later.
             if (furn.objectType == "Door")
             {
-                if (furn.GetParameter("openness") < 0.1f)
+                if (furn.GetParameters()["openness"].ToFloat() < 0.1f)
```
Comment (TomMalbran) [inline Assets/Scripts/Controllers/FurnitureSpriteController.cs]: You should use the property here. And maybe remove the ```GetParameters``` function.

File: Assets/Scripts/Models/Parameter.cs [hunk 2]
```diff
@@ -0,0 +1,202 @@
+﻿using UnityEngine;
```
Comment (TomMalbran) [inline Assets/Scripts/Models/Parameter.cs]: Missing the license header

**Comment by vogonistic** [comment]
Merged!

**Pull request merged**